### ISSUE TITLE
ARCH-4 — Two-phase forward and derivative engine (forces via autograd, stress via strain)

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1563]
**Depends on:** ARCH-3 (#1562) · **Blocks:** ARCH-5 (#1633), FM-00 (#1565), BKD-2 (#1566), DEP-1 (#1583), DEP-2 (#1594), RET-1 (#1597)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** Legacy computes derivative quantities *inside* the model forward: `MACE.forward` calls `prepare_graph` (`mace/modules/utils.py:769-844`) to set `requires_grad` and inject the strain displacement, then `get_outputs` (`utils.py:303-393`) to run `torch.autograd.grad` for forces (`compute_forces`, `utils.py:33`) and virials/stress (`compute_forces_virials`, `utils.py:89`). Those in-forward `autograd.grad` calls are graph breaks that block `torch.compile` adoption. v1 splits this into a **two-phase forward**: the model (`BaseMACE` backbone + `MACEOutputs` output layer, built by ARCH-2 (#1561)/ARCH-3 (#1562)) contains no gradient calls at all, and a **derivative engine** computes forces/stress *around* the model call — the differentiation seam sits outside the (future) compiled graph, which is exactly the region BKD-2 (#1566) later compiles. The known design wrinkle, and why this ticket carries a **design note reviewed before implementation**: the strain displacement must be applied to positions and cell **before** edge vectors are formed (legacy ordering `utils.py:814-830`), so the engine is *not* purely downstream of the backbone — it owns an input-preparation phase before the model call and a differentiation phase after it. Legacy is read in place as the behavioural oracle; reproduce its numbers, never its structure.

**Interface & constraints:**

1. **Two-phase structure.** Phase A (engine, pre): make `positions` — and `displacement` when stress/virials are requested — grad leaves, apply the symmetric strain, form edge vectors, hand the model a ready graph. Phase B (model): `BaseMACE` + `MACEOutputs` forward — no `autograd.grad`, no `requires_grad_` mutation, no host-side branching on what derivatives are wanted. Phase C (engine, post): `torch.autograd.grad` on the model's energy output. Nothing in phase B may prevent `torch.compile(fullgraph=True)` over it later.

2. **Strain math to reproduce exactly** (legacy `get_symmetric_displacement`, `utils.py:118-155`): symmetrize `D_sym = 0.5 * (D + D.transpose(-1,-2))` (`:107-109`); displace positions `pos + einsum("be,bec->bc", pos, D_sym[batch])` (`:110-112`) and cell `cell + cell @ D_sym` (`:113-114`); then **recompute `shifts` from `unit_shifts` and the displaced cell**: `einsum("be,bec->bc", unit_shifts, cell[batch[sender]])` (`:115-119`). Note the displacement leaf is seeded as `zeros + positions.sum() * 0.0` (`:106`) so it is connected to the graph even before the einsums — reproduce the connection, not the idiom. Consequences guaranteed by the graph contract (DATA-3 (#1571)): `unit_shifts` is a **required** key and the authoritative quantity under strain (`shifts` is a derived cache); the displacement is applied **before** edge vectors are computed; `displacement` is an **optional input key** so integrations that manage their own strain keep working (legacy honors `data.get("displacement")` at `utils.py:824`; torchsim supplies it padded, `mace/calculators/mace_torchsim.py:421-425`, built at `:629-635`).

3. **Read-only forward.** The engine returns **new** `(positions, shifts, displacement)` tensors and never writes into the input dict — this deletes the legacy write-back `data["positions"], data["shifts"] = p, s` (`utils.py:825`, repeated in the magnetic model at `mace/modules/extensions.py:1779-1789`), whose caller-visible mutation is a live bug source today.

4. **What is differentiated: `total_energy`, uniformly.** Legacy `ScaleShiftMACE` differentiates the interaction energy only (`mace/modules/models.py:589`, and `MagneticScaleShiftMACE` likewise at `mace/modules/extensions.py:1927`) while legacy `MACE` differentiates the E0-inclusive total (`models.py:404`). With ARCH-3 (#1562)'s readout-only E0s the E0 branch is a table lookup on `node_attrs`/`head` with **no autograd path to `positions` or `displacement`** — `autograd.grad` never traverses it — so `∂total/∂pos ≡ ∂inter/∂pos` bit-for-bit, and the engine unifies both legacy autograd calls at zero numerical cost. One case does **not** follow the rule and must be handled explicitly rather than by extension of it: `MACELES` differentiates `inter_e + les_energy` (`mace/modules/extensions.py:612-613`) because the long-range term is a genuine additional energy contribution with its own position dependence — in v1 that is an extra energy term registered on the output layer, and the engine still differentiates one total. The engine consumes ARCH-3 (#1562)'s typed `EnergyTerms`/`MACEOutput` fields (`total_energy` in model dtype), never raw dicts.

5. **Numeric semantics preserved** (from `compute_forces` / `compute_forces_virials` / `get_outputs`): forces `= −∂E/∂pos`; virials `= −∂E/∂ε` (grad w.r.t. the injected displacement); stress `= virials / |det(cell)|` per graph, masked to zero for a graph with no periodic direction (`cell_volume_and_mask`, `utils.py:52-75`; the mask has to be applied before the division, because a nan routes back through it on the backward pass), with `|stress| < 1e10` kept only as the near-degenerate backstop (`utils.py:78-86`); `allow_unused=True` with `None` gradients mapped to zeros (complete dissociation, `utils.py:110-113`); `create_graph=retain_graph=training` so force training backpropagates through the derivative (the `create_graph=True` second-derivative path P0-6's single-step gradient golden exercises). Edge forces are `−∂E/∂vectors` with the sign flipped once for LAMMPS (`utils.py:381-388`), and `get_atomic_virials_stresses` (`utils.py:394`) derives per-atom virials/stresses from them — both are consumer-facing (DEP-1 (#1583)/DEP-2 (#1594)) but the engine is their only production site. Sign conventions match the fp64 CPU pins from P0-6's finite-difference characterization and the documented convention constants in `mace_core` (forces = −∇E, pinned stress sign) — asserted, never assumed.

6. **Derivation modes.** The observable spec declares how each quantity is derived — `readout | autograd | grad_strain`; derivatives are auto-named `d_<q>_d_pos`, with `energy` special-casing to `forces`/`stress`. The engine is the single execution site for the `autograd` and `grad_strain` modes. **Derivatives with respect to a declared model input other than positions or strain are a fourth mode and belong to ARCH-5 (#1633)**; this ticket only guarantees that adding it needs no change to the phase A/B/C split, and that phase C can request several grad targets in **one** `autograd.grad` call, which is how legacy already computes forces+virials+magforces together (`utils.py:213-259`).

7. **Where it lands:** `packages/mace-torch/src/mace_torch/physics/outputs.py` — the successor of legacy `get_outputs` plus the strain-injection part of `prepare_graph`. No import of legacy `mace/` outside `tests/parity/`; no e3nn.

**Task:**
1. Write the **strain-seam design note** (phase boundaries, strain-before-edge-vectors ordering, read-only contract, differentiate-`total_energy` argument, the multi-target single-`autograd.grad` shape ARCH-5 (#1633) needs, compile-seam placement for BKD-2 (#1566)) and get it reviewed **before** implementing: the engine brackets the backbone rather than sitting purely downstream of it.
2. Implement phase A in `mace_torch/physics/outputs.py`: grad-leaf setup, symmetric strain displacement per the legacy math, `shifts` recomputed from `unit_shifts`, edge vectors formed after displacement, new tensors returned (no dict write-back), optional external `displacement` honored.
3. Implement phase C: forces via `autograd.grad` of `total_energy` w.r.t. positions; virials/stress via the grad w.r.t. displacement; edge forces w.r.t. edge vectors with the LAMMPS sign flip, and per-atom virials/stresses derived from them; legacy conventions, filters, and `create_graph=training` semantics; results populate the typed output fields.
4. Wire the engine to the observable derivation modes (`autograd` / `grad_strain`) so declared derivatives are produced without per-quantity special-case code beyond the `energy`→`forces`/`stress` naming rule.
5. Finite-difference validation reusing the P0-6 methodology on the tiny fixture, fp64 CPU: forces vs 5-point central differences (h≈1e-4 Å, tol ~1e-6 eV/Å); stress vs finite strains (±1e-5), all 6 Voigt components; virial sign convention asserted in the test.
6. Parity tests vs the frozen legacy `mace/` under `tests/parity/`: fp64 energy/forces/stress on the tiny anchors for both legacy model classes, plus an explicit no-input-mutation assertion.

**Out of scope:** the compiled padded path and `PrecisionConfig` (BKD-2 (#1566)); non-positional derivatives, conditioned grad leaves and fixed-point re-entrancy (ARCH-5 (#1633)); Hessians (legacy `compute_hessians_vmap`, `utils.py:159`, and `compute_hessians_loop`, `:155`) and the LAMMPS edge-force consumers — the DEP-1 (#1583) calculator and DEP-2 (#1594) ML-IAP path own those; the ASE calculator itself (DEP-1 (#1583) starts once this lands); the ML-IAP `vectors`-ingestion profile, which replaces positions with edge vectors as the grad leaf entirely (`utils.py:786-803`; deployment adapter, DEP-2 (#1594)); any change to legacy `mace/`.

**Acceptance criteria:**
- [ ] The design note is reviewed and signed off before the implementation PR (seam ordering, read-only contract, differentiation target, multi-target grad shape).
- [ ] Forces match −dE/dx by 5-point central finite differences (h≈1e-4 Å, tol ~1e-6 eV/Å) at fp64 on the tiny fixture.
- [ ] Stress matches (1/V)·dE/dε under finite strains (±1e-5) on all 6 Voigt components; the virial sign convention is asserted, not assumed.
- [ ] fp64 energy/forces/stress parity vs the frozen legacy on the tiny anchors for **both** legacy model classes — confirming that differentiating `total_energy` reproduces both legacy autograd variants bit-for-bit on the position/strain path.
- [ ] No `autograd.grad` and no `requires_grad_` mutation inside any model forward (asserted by test); the engine is the only differentiation site.
- [ ] A stress computation leaves the input dict unchanged (`positions`/`shifts` bit-identical before and after); the engine returns new tensors.
- [ ] An externally supplied `displacement` input key is used instead of a fresh zero displacement (torchsim-style callers keep working).
- [ ] Edge forces reproduce legacy's sign convention, and per-atom virials/stresses derived from them match legacy on the tiny fixture.
- [ ] Model-level force+stress `gradcheck`/`gradgradcheck` pass at fp64 on the tiny fixture (the `create_graph=True` force-training path is second-derivative correct).

**Verify:**
```bash
python -m pytest packages/mace-torch/tests/physics/test_outputs_fd.py -v        # finite differences, sign conventions, gradgradcheck
python -m pytest packages/mace-torch/tests/physics/test_engine_contract.py -v   # read-only dict, no grad calls in forward, displacement passthrough
python -m pytest tests/parity -m "not slow" -k "forces or stress"               # vs frozen legacy mace/
```
**Review focus:** the strain ordering (displacement applied before edge vectors — the one place the engine reaches upstream of the backbone), the virial/stress sign convention (the #1 silent-porting-bug candidate), and the differentiate-`total_energy` equivalence argument together with the `MACELES` exception — numerics-critical, needs a physics reviewer, gated on the pre-implementation design-note review.

---
